### ISSUE TITLE
CRM and initial-input question

### ISSUE THREAD

**Issue by bdarcus** [minad/vertico#59]
Question, @minad:

In vertico (and selectrum), if one specifies initial-input in a CRM interface, the initial-input only applies to the first selection.

Use cases for bibtex-actions:

1. I want to read a collection of keys at-point and pass them to vertico/selectrum to narrow to those three, but select two of them to run the command on.
2. same thing to pre-filter items that have associated PDF flies, or notes

In each case, if one tab-selects a candidate, the initial-input is removed, and the result is the full candidate list.

Another one of those pain points for me in CRM.

Is this something that can/should be fixed in existing CRM UIs, or does it seem like the long term solution is what we've previously discussed at https://github.com/raxod502/selectrum/issues/489, and to allow things like "select all" for a list of such narrowed candidates?

**Comment by minad** [comment]
Maybe I misunderstood the question, but this works for me in both Vertico and Selectrum:

~~~ elisp
(completing-read-multiple "Select: " '("first" "second" "third") nil 'require-match "first,second,")
~~~

**Comment by minad** [comment]
Regarding the completing-read-multiple pain points - currently I don't have plans to "enhance" crm in Vertico. Vertico treats crm as a simple programmable completion table. Selectrum is more intrusive here, since it replaces the existing completion table.

You could experiment with writing your own programmable completion table, take a look at the source of `completing-read-multiple` and `crm--collection-fn`. Maybe you can construct a better completion table which fits your bibtex purposes better. This requires more digging into how these tables are implemented internally. There is one caveat though - programmable (dynamic) completion tables do not work properly in Ivy or Selectrum. They only work in fully compliant UIs like default completion, Icomplete and Vertico.

Basically there are three types of completion tables:

1. Static tables, for example plain lists. Work everywhere.
2. Dynamic/programmable tables. Work only in compliant UIs like default completion, Icomplete or Vertico. Examples: File completion, Crm.
3. Broken tables. Org mode has a few of those (e.g., org-set-tags-command). These tables work only with default completion as long as you do not modify the completion-styles. The tables follow the completion API in a very specific sense and make assumptions about the completion-style, e.g., the basic completion style must be used.

**Comment by hmelman** [comment]
That last part would be good in a README.

**Comment by bdarcus** [comment]
> Maybe I misunderstood the question, but this works for me in both Vertico and Selectrum:
> 
> ```emacs-lisp
> (completing-read-multiple "Select: " '("first" "second" "third") nil 'require-match "first,second,")
> ```

Hi @minad - conceptually, this is what we're doing:

``` emacs-lisp
(completing-read-multiple "Select: " '("a one" "a two" "a three" "b four") nil 'require-match "a")
```

So initially, the first three candidates display.

If I select first and tab, all then display.

So the simple question is why doesn't the initial value get added after the first selection?

But maybe, as you suggest in the subsequent comments, this isn't the best tool for the job. Seems I might want to look into consult asynchronous tables?

**Comment by minad** [comment]
Okay, I understand. This is not how CRM works, but you could write your own CRM-like completion table which adds the initial value every time. Consult async tables are not a solution for the problem here.

**Comment by bdarcus** [comment]
Got it. Thank you!

**Comment by bdarcus** [comment]
> There is one caveat though - programmable (dynamic) completion tables do not work properly in Selectrum.

Just one additional question actually, on this, since it's important: do you see this situation as likely to be resolved? I know there's a PR related to dynamic tables, but not sure how it fits, and I know it's been sitting for awhile.

**Comment by minad** [comment]
> Just one additional question actually, on this, since it's important: do you see this situation as likely to be resolved? I know there's a PR related to dynamic tables, but not sure how it fits, and I know it's been sitting for awhile.

I hope so. It is good to have alternatives and it is good to have different independent packages maintained by different people. With that PR and maybe a few other changes, the issues can be fixed (dynamic tables, completion boundaries). There is another issue on the Selectrum tracker, where the incompatibilities are documented and there have been a few efforts before to get these things sorted out. I also contributed some of the patches or tried to push Selectrum a bit in that direction over the last few months.

Personally, I moved over to Vertico, which is as simple as possible (I believe) and embraces the Emacs filtering mechanism. For this reason Vertico is incompatible with Prescient, since Prescient uses its own filtering API. Vertico started as an experiment on how the dynamic table support could be implemented in Selectrum and then it turned out that this is easy enough to just make it a full completion UI, which is automatically fully compatible. It turns out that the Emacs filtering infrastructure is pretty rigid - you don't have many degrees of freedom to implement a completion UI if you care about full compatibility.
